frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx

```diff
@@ -57,21 +63,29 @@ const ExpressionWidget = (props: ExpressionWidgetProps): JSX.Element => {
 
   const helpTextTargetRef = useRef(null);
 
-  const isValid = !!name && !error && isExpression(expression);
+  const isValidName = withName ? !!name : true;
+  const isValidExpression = !!expression && isExpression(expression);
+
+  const isValid = !error && isValidName && isValidExpression;
 
-  const handleCommit = () => {
+  const handleCommit = (expression: Expression | null) => {
     if (isValid && isNotNull(expression)) {
       onChangeExpression(name, expression);
       onClose && onClose();
     }
   };
 
+  const handleExpressionChange = (parsedExpression: Expression | null) => {
+    setExpression(parsedExpression);
+    setError(null);
+  };
+
   return (
     <Container>
+      {header}
```

Comment:
Can it moved completely outside of this component, so `ExpressionWidget` wouldn't know anything about any header?